### ISSUE TITLE
[QUESTION] end-to-end updates

### ISSUE THREAD

**Issue by davidegraff** [nataliemaus/lolbo#4]
Sorry to bother you again, but I had two questions about the end-to-end retraining from your paper. You write in the paper that LoL-BO introduces a new hyperparameter $\tau_{retrain}$, the frequency with which you perform an end-to-end update of your joint VAE+GP upon successive failures. Given that $\tau_{retrain} << \tau_{fail}$ (trust region update failure frequency), it's possible that you encounter multiple end-to-end updates for every trust region update, but I didn't see any data on this in the paper. Did you look at this / have any rough estimates of how many updates occurred in the `Lol-BO (SELFIES)` traces from figures 2 and 3?

Also, given that you start from a pretrained VAE, are you only ever fine-tuning on **just** the acquired data, or are you retraining on both the acquired data _and_ initial pretraining data

thanks!

**Comment by nataliemaus** [comment]
Indeed it is possible (and likely) that we have multiple end-to-end updates of the joint VAE+GP for every trust region update. In the repo, the total number of times that we have updated the VAE+GP end-to-end is automatically recorded as 'total_number_of_e2e_updates' in Weights and Biases, so you can compare the number of end-to-end updates to the total number of oracle calls, best reward found, current trust region length, etc. during optimization.

**Comment by nataliemaus** [comment]
To address your second question, indeed we do NOT fine-tune the VAE on the entire pre-training dataset since this would be very expensive (i.e. the SELFIES VAE is pre-trained on more than 1 million molecules).

**Comment by davidegraff** [comment]
true- looking at the code though it seems you only ever finetune on the most recent batch and current top-`k` molecules ([L167-173](https://github.com/nataliemaus/lolbo/blob/main/lolbo/lolbo.py#L167)). Given that all of your optimizations start with an initialization dataset of ~10k molecules, the e2e finetuning could "reasonably" be done with all _acquired_ molecules. However, if I'm tracing the code correctly, it seems that this finetuning only ever happens on ~1000 molecules (assuming you used default settings for all molecular optimization experiments). Is this understanding correct? If so, did you ever experiment with larger finetuning batches? thanks!

**Comment by nataliemaus** [comment]
Your understanding is correct, and indeed you could fine-tune on all acquired molecules instead of just the top-1k. However, we briefly tried this and we found that fine-tuning on the full set of >10k molecules acquired did not improve performance, and made optimization runs take significantly longer since forwarding through the large VAE model is expensive.

**Comment by davidegraff** [comment]
Good to know! We were looking at something similar but the cost of the frequent VAE updates scared us away, so it's pretty cool that you can realize significant benefits with such a _cheap_ update. Given that the parameter `k` controls a large amount of the retraining dynamics, did you experiment with different values?

**Comment by nataliemaus** [comment]
Yes, we did do some experimentation since we wanted to use k as small as possible to minimize computation cost, but large enough that performance does not degrade. We found that for guacamole tasks, using k = 1000 is ideal since performance begins to degrade with k<1000. If you are planning on extending LOL-BO to new tasks I recommend doing a similar hyperparameter search to find an ideal value of k for your task. Hope that helps!
